node/src/reactor/main_reactor/keep_up.rs

```diff
@@ -681,34 +692,60 @@ impl MainReactor {
             }
         }
     }
+}
 
-    fn synced_to_ttl(&self) -> Result<bool, String> {
-        if let Some(latest_switch_block_header) = self
-            .storage
-            .read_highest_switch_block_headers(1)
-            .map_err(|err| err.to_string())?
-            .last()
-        {
-            let maybe_highest_orphaned_block_header =
-                self.storage.get_highest_orphaned_block_header();
+pub(crate) fn synced_to_ttl(
+    maybe_latest_switch_block_header: Option<&BlockHeader>,
+    maybe_highest_orphaned_block_header: Option<&BlockHeader>,
+    storage: &Storage,
+    max_ttl: TimeDiff,
+) -> Result<bool, String> {
+    let switch_block_header = match maybe_latest_switch_block_header {
+        Some(switch_block_header) => switch_block_header.clone(),
+        None => {
+            // If latest switch block header is not provided, we try to get it from storage.
+            if let Some(latest_switch_block_header) = storage
+                .read_highest_switch_block_headers(1)
+                .map_err(|err| err.to_string())?
+                .last()
+            {
+                latest_switch_block_header.clone()
+            } else {
+                // No latest switch block header is known, we assume that we're not synced to
+                // TTL.
+                info!(
+                    "KeepUp: unable to get latest switch block header (this problem should be transient), \
+                    assuming not synced to ttl");
+                return Ok(false);
+            }
+        }
+    };
```

Comment:
I would like to see this logic extracted to either in the outer calling function or in a separate function. Same for the other block we should have. The reason is that I think we shouldn't be passing storage around like that and it would be more modular, but it's ultimately up to you. In case we can't get either the switch block or orphaned block, we can exit early in the calling functions.

Reply:
It'll be addressed here: https://github.com/casper-network/casper-node/pull/3868